src/contexts/session.context.tsx

```diff
@@ -0,0 +1,58 @@
+import { createContext, PropsWithChildren, useContext, useState } from 'react';
+import { Blockchain } from '../api/definitions/blockchain';
+import { ApiError } from '../api/definitions/error';
+import { useApiSession } from '../api/hooks/api-session.hook';
+import { useWalletContext } from './wallet.context';
+
+export interface SessionInterface {
+  isMetaMaskInstalled: boolean;
+  address?: string;
+  blockchain?: Blockchain;
+  isLoggedIn: boolean;
+  needsSignUp: boolean;
+  login: () => Promise<void>;
+  signUp: () => Promise<void>;
+}
+
+const SessionContext = createContext<SessionInterface>(undefined as any);
+
+export function useSessionContext(): SessionInterface {
+  return useContext(SessionContext);
+}
+
+export function SessionContextProvider(props: PropsWithChildren): JSX.Element {
+  const { isLoggedIn, getSignMessage, createSession } = useApiSession();
+  const { isInstalled, isConnected, address, blockchain, connect, signMessage } = useWalletContext();
+  const [needsSignUp, setNeedsSignUp] = useState(false);
+  const [signature, setSignature] = useState<string>();
```

Comment:
Theoretically, this could be solved without state, when using a callback method. In case of 404, the login method calls the callback method (which will trigger the UI to show the TNC). After successful completion, the login method continues with sign up.

Reply:
I don't know if it makes it easier. I was just a bit confused by this `needsSignUp` state.

Reply:
yes, but I thought this would be more readable. If it is okay, I will write is an extra improvement ticket

Reply:
created ticket